### PR TITLE
rhvoice: init at 0.5

### PR DESCRIPTION
**Pull request by berce**
###### Motivation for this change

RHVoice is one of the best speech synthesis engines I 've heard, including commercial ones.
It can be tested with `echo "Talk to me" | RHVoice-test`.

###### Things done

- [x] Win the battle against `scons` to gain power over the environment.

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/audio/rhvoice/default.nix [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  # SConstruct patches
+  # 1. Scons creates an independent environment that assumes standard POSIX paths.
+  #    The patch is needed to push the nix environment.
+  #    - PATH
+  #    - PKG_CONFIG_PATH, to find available (sound) libraries
+  #    - RPATH, to link to the newly built libraries
+  # 2. Some data is being zipped. Zip can't handle files from 1970.
+  #
+  postPatch = ''
```
Comment (jtojnar) [inline pkgs/applications/audio/rhvoice/default.nix]: Could you make this a proper patch?

File: pkgs/applications/audio/rhvoice/default.nix [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+        env.PrependENVPath("PATH", os.environ["PATH"])
+        env["ENV"]["PKG_CONFIG_PATH"]=os.environ["PKG_CONFIG_PATH"]
+            '
+    substituteInPlace SConstruct --replace 'else:
+            env["BUILDDIR"]=BUILDDIR
+    ' 'else:
+            env["BUILDDIR"]=BUILDDIR
+            env["RPATH"]="'$out'/lib"
+    '
+    find "./" '!' -newermt '1980-01-01' -exec touch -d '1980-01-02' '{}' '+'
```
Comment (jtojnar) [inline pkgs/applications/audio/rhvoice/default.nix]: Why is this necessary?

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/applications/audio/rhvoice/default.nix [hunk 3]
```diff
@@ -0,0 +1,62 @@
+{ stdenv, lib, pkgconfig, fetchFromGitHub, scons, python, glibmm, libao
+}:
+
+with lib;
```
[the remark this answers is not in the document]
Comment (berce) [reply pkgs/applications/audio/rhvoice/default.nix]: In the meta attribute.
Moved.

File: pkgs/applications/audio/rhvoice/default.nix [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+stdenv.mkDerivation rec {
+  name = "rhvoice";
```
[the remark this answers is not in the document]
Comment (berce) [reply pkgs/applications/audio/rhvoice/default.nix]: Added

File: pkgs/applications/audio/rhvoice/default.nix [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+  #version = "2017-09-24";
```
[the remark this answers is not in the document]
Comment (berce) [reply pkgs/applications/audio/rhvoice/default.nix]: Removed.
I hope to update to that version soon though, 0.5 is several years old.

File: pkgs/applications/audio/rhvoice/default.nix (hunk 2 above)
Comment (berce) [reply pkgs/applications/audio/rhvoice/default.nix]: 2. Some data is being zipped. Zip can't handle files from 1970.

**Comment by grahamc** [comment]
@GrahamcOfBorg eval

(sorry for the noise, master was broken by a merge)

File: pkgs/applications/audio/rhvoice/default.nix (hunk 1 above)
Comment (berce) [reply pkgs/applications/audio/rhvoice/default.nix]: Done.

File: pkgs/development/libraries/speechd/default.nix [hunk 6]
```diff
@@ -62,9 +66,15 @@ in stdenv.mkDerivation rec {
...
   ;
 
+  patches = [ ./add_rhvoice_module.patch ./set_nix_defaultModule.patch ];
+
   postPatch = ''
-    substituteInPlace config/speechd.conf --replace "DefaultModule espeak" "DefaultModule ${selectedDefaultModule}"
+    substituteInPlace config/speechd.conf --replace "nixDefault" "${selectedDefaultModule}"
+    substituteInPlace config/speechd.conf --replace "sd_rhvoice" "${rhvoice}/bin/sd_rhvoice"
+    substituteInPlace config/speechd.conf --replace "RHVoice.conf" "${rhvoice}/etc/RHVoice/RHVoice.conf"
+    ${stdenv.lib.optionalString withPico ''
```
Comment (jtojnar) [inline pkgs/development/libraries/speechd/default.nix]: This conditional is not needed since the svox dependency can be compiled out. RHVoice, on the other hand, cannot so that one should get a conditional. Though, I would move enabling the rhvoice to a NixOS module.

Comment (berce) [reply pkgs/development/libraries/speechd/default.nix]: I rearranged the optional stuff in patches and postPatch. Is that what you wanted?

Since speech-dispatcher ships with several speech engines, I think including RHVoice in the same way is convenient for the user. Would a NixOS module not make it only available in NixOS?

**Comment by volth** [comment]
@GrahamcOfBorg eval

**Comment by jtojnar** [comment]
I remain unconvinced that rhvoice should be part of the speechd expression.

**Comment by volth** [comment]
> I remain unconvinced that ```rhvoice``` should be part of the ```speechd``` expression.

We can separate the PR to 2: adding ```rhvoice``` (no objection here?) and integration with ```speechd``` (awaiting for more feedback; I do not need the integration too although have nothing against it)

**Comment by jtojnar** [comment]
Right, I am fine with `rhvoice` as is, but third-party speech modules should be enabled through NixOS module.

**Comment by volth** [comment]
I updated ```0.5``` to current git (```0.5``` is from 2014) and remove speechd integration:
https://github.com/berce/nixpkgs/pull/1
https://github.com/NixOS/nixpkgs/compare/master...volth:upstream-rhvoice2

**Comment by jtojnar** [comment]
Merged the updated rhvoice as https://github.com/NixOS/nixpkgs/commit/f97c4668abb630875318f057d452f8ebfe52f8e2

**Comment by c0bw3b** [comment]
@jtojnar should we close this since #37370 is merged?
A speechd module for rhvoice will probably go through a new PR anyway
